addons/payment_toss_payments/controllers/main.py

```diff
@@ -0,0 +1,127 @@
+from werkzeug.exceptions import BadRequest, Forbidden
+
+from odoo import http
+from odoo.exceptions import ValidationError
+from odoo.http import request
+
+from odoo.addons.payment.logging import get_payment_logger
+
+_logger = get_payment_logger(__name__)
+
+
+class TossPaymentsController(http.Controller):
+    _success_url = '/payment/toss-payments/success'
+    _fail_url = '/payment/toss-payments/fail'
+    _webhook_url = '/payment/toss-payments/webhook'
+
+    @http.route(_success_url, type='http', auth='public')
+    def handleValidationSuccess(self, **payment_data):
+        """ Process payment after validation from Toss SDK is successful.
+
+        This controller endpoint should be accessed via Toss SDK's successUrl redirection. The
+        method handles payment using Toss API.
+
+        :param dict payment_data: The dict data of successUrl's query parameter.
+        Expected keys = ["orderId", "paymentKey", "amount"]
+        """
+        tx_sudo = request.env['payment.transaction'].sudo() \
+            ._search_by_reference('toss_payments', payment_data)
+
+        if not tx_sudo:
+            _logger.error("Cannot find matching payment trasaction using reference: %s",
+                          payment_data.get('orderId'))
+            return request.redirect('/payment/status')
+
+        try:
```

Comment:
Before making the payment request, you could verify the keys in your payment_data.

If code/message -> FailingURL:
 directly call `_set_error` with the message you want to log (could be a formatting of code + message + orderId). 

If paymentKey/amount-> SuccessURL:
keep going with the amount check and payment request. 

That way, you simplify the code while keeping the distinction between the received data in case of fail/success.